ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs

```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
-      testProperty "validGenerator"                   prop_validGenerator
-    , testProperty "appendPreservesInvariant"         prop_appendPreservesInvariant
-    , testProperty "rewindPreservesInvariant"         prop_rewindPreservesInvariant
-    , testProperty "rewindReappendId"                 prop_rewindReappendId
-    , testProperty "appendOldStatePreservesInvariant" prop_appendOldStatePreservesInvariant
-    , testProperty "appendOldStateRestoresPreWindow"  prop_appendOldStateRestoresPreWindow
+tests = testGroup "PBftChainState" $
+    -- these tests are cheap, so inflate the count
+    map (adjustOption $ \(QuickCheckTests n) -> QuickCheckTests $ 10 * n) $
+    [ testProperty "validGenerator"
+               prop_validGenerator
+    , testProperty "directABb"
+               prop_directABb
+    , testProperty "appendIsMonoidAction"
+               prop_appendIsMonoidAction
+    , testProperty "appendPreservesInvariant"
+               prop_appendPreservesInvariant
+    , testProperty "rewindPreservesInvariant"
+               prop_rewindPreservesInvariant
+    , testProperty "rewindReappendId"
+               prop_rewindReappendId
+    , testProperty "appendOldStatePreservesInvariant"
+               prop_appendOldStatePreservesInvariant
+    , testProperty "appendOldStateRestoresPreWindow"
+               prop_appendOldStateRestoresPreWindow
     ]
 
 {-------------------------------------------------------------------------------
   Test setup
 -------------------------------------------------------------------------------}
 
+-- | A generated test scenario
+--
+-- We don't want to make too many assumptions here, so instead we model what
+-- happens on an actual chain: we build chain states by starting from genesis
+-- and iterating amongst the following actions.
+--
+--   * Append an actual block signed by a key, via 'CS.append'.
+--   * Append an epoch boundary block (EBB), via 'CS.appendEBB'.
+--   * Rewind the chain state to a previous slot, via 'CS.rewind'.
+--
+-- A rewind is expected to fail in two scenarios.
+--
+--   * When the slot is so far back that we would be discarding more than @k@
+--     signed blocks.
+--
+--   * When no block (neither actual nor EBB) was appended in that exact slot.
+--
+-- Rewinding to a slot @s2@ that only contained an EBB should behave similarly
+-- to rewinding to the latest /signed slot/ @s1 < s2@ or to 'Origin' if there
+-- isn't one. More specifically, it should discard the same signed blocks
+-- (those in slots @> s1@) but still retain information about the subsequent
+-- EBBs in slots in the inclusive range from @s1@ to @s2@.
+--
+-- NOTE: PBftChainState assumes @k >= 1@.
+--
+-- * 'testChainInputsA', 'testChainInputsB', and 'testChainInputsC'
+--
+-- We first generate three segments of 'Inputs'.
+--
+-- >                  B
+-- >              ---------
+-- >       A     /
+-- >   ---------<
+-- >             \
+-- >              ---------
+-- >                  C
```

Comment:
Can you draw `C` longer than `B`, as it will contain at least `n` more blocks.

Reply:
Fixed, thanks.